File: python-code/programmers/L2/귤 고르기.py
import heapq


def solution(k, tangerine):
    count_dict = dict()
    for ele in tangerine:
        if ele not in count_dict:
            count_dict[ele] = 0
        count_dict[ele] += 1
    heap = []
    for v in count_dict.values():
        heapq.heappush(heap, -v)
    count = 0
    while True:
        count += 1
        val = abs(heapq.heappop(heap))
        if val >= k:
            break
        k -= val
    return count


print(solution(6,	[1, 3, 2, 5, 4, 5, 2, 3]))  # 3
print(solution(4,	[1, 3, 2, 5, 4, 5, 2, 3]))  # 2
print(solution(2,	[1, 1, 1, 1, 2, 2, 2, 3]))  # 1

File: python-code/programmers/L2/귤 고르기.js
function solution(k, tangerine) {
  var answer = 0
  const count_map = {}
  tangerine.forEach((ele) => {
    if (count_map[ele] == null) count_map[ele] = 0
    count_map[ele]++
  })

  // const sorted_count_list = Object.entries(count_map)
  // sorted_count_list.sort((a, b) => b.at(1) - a.at(1))
  // for (const [_, v] of sorted_count_list) {
  //     if (k <= 0) break
  //     answer++
  //     k -= v
  // }

  const sorted_value_list = Object.values(count_map)
  sorted_value_list.sort((a, b) => b - a)
  for (const count of sorted_value_list) {
    if (k <= 0) break
    answer++
    k -= count
  }
  return answer
}
